solidity/contracts/system/KeepFactorySelection.sol

```diff
@@ -150,29 +170,80 @@ library KeepFactorySelection {
+        } else {
+            IBondedECDSAKeepFactory keepStakeFactory = IBondedECDSAKeepFactory(
+                _self.keepStakeVendor.selectFactory()
+            );
+
+            require(
+                address(keepStakeFactory) != address(0),
+                "Vendor returned invalid factory address"
+            );
+
+            return keepStakeFactory;
+        }
+    }
+
+    /// @notice Returns ETH-stake based factory address. If factories lock is not set
+    /// it calls the ETH staked vendor to obtain the latest versions of the factory.
+    function getFullyBackedFactory(Storage storage _self)
+        internal
+        view
+        returns (IBondedECDSAKeepFactory)
+    {
+        if (_self.factoriesVersionsLock) {
+            return _self.ethStakeFactory;
+        } else {
+            IBondedECDSAKeepFactory ethStakeFactory = IBondedECDSAKeepFactory(
+                _self.ethStakeVendor.selectFactory()
+            );
+
+            require(
+                address(ethStakeFactory) != address(0),
+                "Vendor returned invalid factory address"
+            );
+
+            return ethStakeFactory;
+        }
+    }
+
     /// @notice Sets the address of the fully backed, ETH-stake based keep
     /// factory. KeepFactorySelection can work without the fully-backed keep
     /// factory set, always selecting the default KEEP-stake-based factory.
     /// Once both fully-backed keep factory and factory selection strategy are
     /// set, KEEP-stake-based factory is no longer the default choice and it is
     /// up to the selection strategy to decide which factory should be chosen.
     /// @dev Can be called only one time!
-    /// @param _fullyBackedFactory Address of the fully-backed, ETH-stake based
-    /// keep factory.
-    function setFullyBackedKeepFactory(
+    /// @param _fullyBackedVendor Address of the fully-backed, ETH-stake based
+    /// keep vendor.
+    function setFullyBackedKeepVendor(
         Storage storage _self,
-        address _fullyBackedFactory
+        address _fullyBackedVendor
     ) internal {
         require(
-            address(_self.ethStakeFactory) == address(0),
-            "Fully backed factory already set"
+            address(_self.ethStakeVendor) == address(0),
+            "Fully backed vendor already set"
         );
-        require(
-            address(_fullyBackedFactory) != address(0),
-            "Invalid address"
+        require(_fullyBackedVendor != address(0), "Invalid address");
+
+        IBondedECDSAKeepVendor ethStakeVendor = IBondedECDSAKeepVendor(
+            _fullyBackedVendor
         );
+        address ethStakeFactory = ethStakeVendor.selectFactory();
 
-        _self.ethStakeFactory = IBondedECDSAKeepFactory(_fullyBackedFactory);
+        require(ethStakeFactory != address(0), "Vendor returned invalid factory address");
+
+        _self.ethStakeVendor = ethStakeVendor;
+        _self.ethStakeFactory = IBondedECDSAKeepFactory(ethStakeFactory);
```

Comment:
It probably makes sense to check whether the vendor returns non-zero factory but I think we shouldn't write the factory here. It's not the responsibility of this function.

Reply:
86197d5831142ed319553997437a898b36d31b89, f3137845cb96e549f647819a64f96844fad69014